File: src/main/java/io/protobase/f7/models/ColumnRowKey.java
package io.protobase.f7.models;

import com.google.common.base.MoreObjects;

/**
 * Represents a single location in a 2D grid - usually {@link Grid}.
 */
public class ColumnRowKey extends BaseObject {
  private Integer column;
  private Integer row;

  /**
   * Create a new key from 0-indexed column and 0-indexed row.
   *
   * @param column int greater than or equal to 0.
   * @param row    int greater than or equal to 0.
   */
  public ColumnRowKey(Integer column, Integer row) {
    this.column = column;
    this.row = row;
  }

  /**
   * Column index.
   *
   * @return column index.
   */
  public Integer getColumnIndex() {
    return column;
  }

  /**
   * Row index.
   *
   * @return row index.
   */
  public Integer getRowIndex() {
    return row;
  }

  @Override
  public String toString() {
    return MoreObjects.toStringHelper(this)
        .add("column", column)
        .add("row", row)
        .toString();
  }

  @Override
  public Object[] significantAttributes() {
    return new Object[]{
        column,
        row
    };
  }

  public int compareTo(ColumnRowKey other) {
    if (this.equals(other)) {
      return 0;
    }
    if (this.column.equals(other.getColumnIndex())) {
      return this.row.compareTo(other.getRowIndex());
    }
    return this.column.compareTo(other.getColumnIndex());
  }
}

File: src/main/js/models/common/ColumnRowKey.ts
import { AlphaUtils } from "../../utils/AlphaUtils";
import { Compare } from "../../utils/Compare";
import { A1Key } from "./A1Key";
import { Hashable } from "../../common/standard/Hashable";

/**
 * Represents a single location in a 2D grid - usually {@link Grid}.
 */
export class ColumnRowKey implements Hashable {
  readonly column: number;
  readonly row: number;

  constructor(column: number, row: number) {
    this.column = column;
    this.row = row;
  }

  get hashKey(): string {
    return JSON.stringify([this.column, this.row]);
  }

  static fromA1(a1: string) {
    const key = A1Key.fromString(a1);
    return new ColumnRowKey(key.getColumnIndex(), key.row);
  }

  toA1(): string {
    return `${AlphaUtils.oneIndexedNumberToLetter(this.column + 1)}${this.row + 1}`;
  }

  compareTo(other: ColumnRowKey): number {
    if (this.row === other.row && this.column === other.column) {
      return 0;
    }
    if (this.column === other.column) {
      return Compare.numberComparison(this.row, other.row);
    }
    return Compare.numberComparison(this.column, other.column);
  }

  equals(other: unknown): boolean {
    return other instanceof ColumnRowKey && this.hashKey === other.hashKey;
  }
}
